spark/src/test/scala/org/apache/spark/sql/delta/DeltaProtocolVersionSuite.scala

```diff
@@ -575,17 +582,203 @@ trait DeltaProtocolVersionSuiteBase extends QueryTest
+
+      validateUnsupportedTableReadFeatureException(
+        exceptionRead,
+        pathInErrorMessage,
+        invalidReaderFeatures)
+
+      version = version + 1
+      val invalidWriterFeatures = Seq("NonExistingWriterFeature1", "NonExistingWriterFeature2")
+      val protocolWriterFeatures = Protocol(
+        TABLE_FEATURES_MIN_READER_VERSION,
+        TABLE_FEATURES_MIN_WRITER_VERSION)
+        .withWriterFeatures(invalidWriterFeatures)
+      untrackedChangeProtocolVersion(deltaLog, version, protocolWriterFeatures)
+      if (!warm) {
+        DeltaLog.clearCache()
+      }
+
+      val exceptionWrite = intercept[DeltaUnsupportedTableFeatureException] {
+        spark.range(1).write
+          .mode("append")
+          .option("mergeSchema", "true")
+          .format("delta")
+          .saveAsTable(featureTable)
+      }
+
+      validateUnsupportedTableWriteFeatureException(
+        exceptionWrite,
+        pathInErrorMessage,
+        invalidWriterFeatures)
+
+      // Restore the protocol version or the clean-up fails
+      version = version + 1
+      untrackedChangeProtocolVersion(deltaLog, version, 1, 2)
+    }
+  }
+
+  test("DeltaUnsupportedTableFeatureException - error message with table name - warm") {
+    testTableFeatureErrorMessageWithTableName(warm = true)
+  }
+
+  test("DeltaUnsupportedTableFeatureException - error message with table name - cold") {
+    testTableFeatureErrorMessageWithTableName(warm = false)
+  }
+
+  test("DeltaUnsupportedTableFeatureException - " +
+    "incompatible protocol change during the transaction - table name") {
+    for ((incompatibleProtocol, read) <- Seq(
+        (Protocol(
+          TABLE_FEATURES_MIN_READER_VERSION,
+          TABLE_FEATURES_MIN_WRITER_VERSION)
+          .withReaderFeatures(Seq("NonExistingReaderFeature1", "NonExistingReaderFeature2")),
+          true),
+        (Protocol(
+          TABLE_FEATURES_MIN_READER_VERSION,
+          TABLE_FEATURES_MIN_WRITER_VERSION)
+          .withWriterFeatures(Seq("NonExistingWriterFeature1", "NonExistingWriterFeature2")),
+          false)
+    )) {
+      val tableName = "mytablefeaturesnotsupported"
+      withTable(tableName) {
+        spark.range(0).write.format("delta").saveAsTable(tableName)
+        val deltaLog = DeltaLog.forTable(spark, TableIdentifier(tableName))
+        val catalogTable = DeltaTableV2(spark, TableIdentifier(tableName)).catalogTable
+        val txn = deltaLog.startTransaction(catalogTable)
+        val currentVersion = txn.snapshot.version
+        untrackedChangeProtocolVersion(deltaLog, currentVersion + 1, incompatibleProtocol)
+
+        // Should detect the above incompatible feature and fail
+        val exception = intercept[DeltaUnsupportedTableFeatureException] {
+          txn.commit(AddFile("test", Map.empty, 1, 1, dataChange = true) :: Nil, ManualUpdate)
+        }
+
+        var pathInErrorMessage = "default." + tableName
+
+        read match {
+          case true =>
```

Comment:
It was just because I initially tried to do the same as we do in Delta Log to create unsupportedFeaturesException pointer:

https://github.com/delta-io/delta/blob/60e187a34f1bd8f8f5a081480427850f5628e129/spark/src/main/scala/org/apache/spark/sql/delta/DeltaLog.scala#L373C1-L382C6

Then realized the last parameter is different and that wouldn't work, but forgot to change this to if/else

Reply:
Let me know if you prefer me to change, I think the code is more clear using if/else

Reply:
Let's leave it for future cleanup -- I'd hate to delay the merge yet again over this.